Bug fix in util.misc.run_in_new_terminal

beofek4 opened the pull request:
Bug:
when cmd is a string, terminator is failed to open with the error "Unable to find a shell"

# Pwntools Pull Request

Thanks for contributing to Pwntools!  Take a moment to look at [`CONTRIBUTING.md`][contributing] to make sure you're familiar with Pwntools development.

Please provide a high-level explanation of what this pull request is for.

## Testing

Pull Requests that introduce new code should try to add doctests for that code.  See [`TESTING.md`][testing] for more information.

## Target Branch

Depending on what the PR is for, it needs to target a different branch.

You can always [change the branch][change] after you create the PR if it's against the wrong branch.

| Branch   | Type of PR                                                       |
| -------- | ---------------------------------------------------------------- |
| `dev`    | New features, and enhancements
| `dev`    | Documentation fixes and new tests
| `stable` | Bug fixes that affect the current `stable` branch
| `beta`   | Bug fixes that affect the current `beta` branch, but not `stable`
| `dev`    | Bug fixes for code that has never been released

[contributing]: https://github.com/Gallopsled/pwntools/blob/dev/CONTRIBUTING.md
[testing]: https://github.com/Gallopsled/pwntools/blob/dev/TESTING.md
[change]: https://github.com/blog/2224-change-the-base-branch-of-a-pull-request

beofek4 committed:
Bug fix in util.misc.run_in_new_terminal

Bug:
when cmd is a string terminator is failed to open with the error "Unable to find a shell"

Fix:
I spilted the command to a list

zachriggle commented:
Please add a test that demonstrates the fix, so we can be sure not to re-introduce it.

beofek4 committed:
Add test and fix some bugs

I can't check if the terminal as opened with the error "Unable to find a shell"
so I just checked if the function returned a number, the tester should check if a terminal has been opened successfully.

* the test should be with terminator as the defualt terminal.

beofek4 commented:
I added a test and I noticed that a normal split isn't good enough,
if you want one of the arguments will be with a space it will not work as we want.
so I used shlex.split instead.

beofek4 committed:
Merge branch 'stable' into stable

zachriggle commented on pwnlib/util/misc.py:
```diff
@@ -201,6 +202,14 @@ def run_in_new_terminal(command, terminal = None, args = None):
+
+    Examples:
+        >>> type(run_in_new_terminal('/usr/bin/gdb -q /bin/ls'))
+        <type 'int'>
```
This is an insufficient test.  You need to run something that generates output that can be checked.

beofek4 replied:
run_in_new_terminal opens a new terminal and writes to it, I am not sure how to check it from doctest, do you have any idea of how can I do that?

zachriggle replied:
Something that touches a temporary file would work.

```py
>>> path = tempfile.mktemp()
>>> pid = run_in_new_terminal(['rm', '-f', path])
>>> pid, status = os.waitpid(pid)
>>> status == 0
True
>>> os.path.exists(path)
False
```

zachriggle replied:
Updated my example to be a bit more robust

zachriggle commented on pwnlib/util/misc.py:
```diff
@@ -216,6 +225,8 @@ def run_in_new_terminal(command, terminal = None, args = None):
             terminal = 'x-terminal-emulator'
             args     = ['-e']
+            if isinstance(command, str):
+                command  = shlex.split(command)
```
String splitting is a convenience for basic commands.  

`shlex` doesn't follow the same string-splitting as *every* shell, and does not understand most Unix escapes.

If your command is complicated you **MUST** use the `list` of arguments.  Perhaps we should deprecate string splitting altogether?

beofek4 replied:
I thought maybe to mimic system and use /bin/bash to split the args,
command = ["/bin/bash", "-c", command].
do you think it is a good idea?

zachriggle replied:
You'll want to use `/bin/sh`, but that could work, sure.

zachriggle commented on pwnlib/util/misc.py:
```diff
@@ -359,3 +370,8 @@ def register_sizes(regs, in_sizes):
     return lists.concat(regs), sizes, bigger, smaller
+
+
+if __name__ == "__main__":
```
This block is unnecessary, please remove it.  See `TESTING.md` for how to run tests.

beofek4 committed:
Improve test and change spliting method

I changed the splitting method, now it is mimic system and using /bin/sh
I changed the test to better test that check if the function executes the command or not.

beofek4 committed:
Merge branch 'stable' of https://github.com/beofek4/pwntools into stable

beofek4 committed:
Merge branch 'stable' of https://github.com/beofek4/pwntools into stable

beofek4 committed:
Merge branch 'stable' into stable

Arusekk committed:
Merge branch 'stable' into stable

Arusekk commented:
I believe `context.terminal` is not set anywhere, so the code can't guess a terminal type in Travis.

zachriggle commented:
We don’t test the code on Travis IIRC

On Mon, Dec 9, 2019 at 11:01 AM Arusekk <notifications@github.com> wrote:

> I believe context.terminal is not set anywhere, so the code can't guess a
> terminal type in Travis.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Gallopsled/pwntools/pull/1020?email_source=notifications&email_token=AAA3IGCHULB2Z7EJTJVJZKLQXZ2VRA5CNFSM4DZVXY3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGJ4UQI#issuecomment-563333697>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3IGAA5KCPPQ6GSFUPXYLQXZ2VRANCNFSM4DZVXY3A>
> .
>
-- 

*Zach Riggle*

heapcrash commented:
Closing this out since the PR is two+ years old.  Please rebase your changes on top of the current `dev` branch or create a new PR to get this merged.

dp12 commented:
If anyone is having this issue with terminator like me, I found from a Chinese Q&A site that adding the following line fixes the issue:

    context.terminal = ['terminator','-x','sh','-c']
